src/FeedGenerator.php

```diff
@@ -629,39 +710,36 @@ public function handle_action_timeout( $action_id, $error ) {
 	 * @since 1.2.14
 	 */
 	protected function is_timeout_error( array $error ): bool {
-		return isset( $error['type'] ) &&
-				E_ERROR === $error['type'] &&
-				isset( $error['message'] ) &&
-				strpos( $error ['message'], 'Maximum execution time' ) !== false;
+		$is_error              = E_ERROR === $error['type'] ?? 0;
+		$is_max_execution_time = strpos( $error['message'] ?? '', 'Maximum execution time' ) !== false;
+		return $is_error && $is_max_execution_time;
 	}
 
 	/**
-	 * Handle error on generate feed timeout.
+	 * Check whether the action's failure rate is above the specified threshold within the timeframe.
 	 *
-	 * @since 1.2.14
+	 * @param string $hook The job action hook.
+	 * @param ?array $args The job arguments.
 	 *
-	 * @param int $action_id The ID of the action marked as failed.
+	 * @return bool True if the action's error rate is above the threshold, and false otherwise.
 	 *
-	 * @throws Exception Related to max retries reached or missing arguments on the action.
+	 * @since x.x.x
 	 */
-	public function maybe_handle_error_on_timeout( int $action_id ) {
-
-		$action = ActionScheduler::store()->fetch_action( $action_id );
-
-		if ( $this->get_action_full_name( self::CHAIN_BATCH ) !== $action->get_hook() ) {
-			return;
-		}
-
-		try {
-			$action_args = $action->get_args();
-
-			if ( ! isset( $action_args[0] ) || ! is_int( $action_args[0] ) || ! isset( $action_args[1] ) || ! is_array( $action_args[1] ) ) {
-				throw new Exception( __( 'There was not possible to re-schedule the action, no args available.', 'pinterest-for-woocommerce' ) );
-			}
+	protected function is_failure_rate_above_threshold( string $hook, ?array $args = null ): bool {
+		$threshold   = apply_filters( 'pinterest_for_woocommerce_action_failure_threshold', 3 );
+		$time_period = apply_filters( 'pinterest_for_woocommerce_action_failure_time_period', 1800 ); // 30 minutes.
+		$failed_actions = $this->action_scheduler->search(
+			[
+				'hook'         => $hook,
+				'args'         => $args,
+				'status'       => ActionSchedulerInterface::STATUS_FAILED,
+				'per_page'     => $threshold,
```

Comment:
I've been giving this a bit more thought, and I think the combination of `'per_page' => $threshold` and the `count()` comparison below could use an adjustment. I don't think it won't work, but I do think it can be a bit challenging to be certain it's working as we expect.

I think we can safely eliminate the `per_page` parameter here, and then go back to the `>=` comparison in the `return` statement. This query is still bound by a date, so we're not at risk of querying too many actions and getting a higher number than we expect.

Reply:
Do we want to control the failure rate, let's say we don't want action to retry after two or four attempts (three by. default) and `per_page` will help us to do that when provided with a value from the `pinterest_for_woocommerce_action_failure_threshold ` filter. Unless we just do it up to three times and give up. 

I am still not sure if it is reasonable to have that level of control on the number of retries though.

I think we have two options here: 
- leave per_page and have a filter to redefine the number of attempts.
- remove per_page, remove the filter and have three attempts hardcoded.

What do you think, Jeremy?